File: backend/app/services/executive_dashboard.py
import asyncio
import time
from datetime import datetime, timezone
from decimal import Decimal
from typing import Optional

from app.core.config import settings
from app.core.permissions import user_has_permission
from app.models.user import User
from app.repositories.app_metrics import AppMetricsRepository
from app.repositories.customer import CustomerRepository
from app.repositories.digital_metal_inventory import DigitalMetalInventoryRepository
from app.repositories.report import ReportRepository
from app.repositories.user import UserRepository
from app.repositories.workflow import WorkflowRepository
from app.schemas.dashboard import (
    AppDashboardMetrics,
    AssignedTaskSummary,
    CustomerDashboardMetrics,
    DailyActivityItem,
    ExecutiveDashboardResponse,
    ExecutiveRole,
    InventoryDashboardMetrics,
    RevenueTrendPoint,
    TeamDashboardMetrics,
    TransactionDashboardMetrics,
    WorkflowApprovalSummary,
)
from app.schemas.digital_metal_inventory import compute_stock_status
from app.schemas.inventory import InventoryItemResponse
from app.services.audit import AuditService
from app.services.inventory import InventoryService
from app.services.metal_prices import MetalPriceService
from app.services.notification import NotificationService
from app.services.transaction import TransactionService

_executive_cache: dict[str, tuple[float, ExecutiveDashboardResponse]] = {}


def resolve_executive_role(user: User) -> ExecutiveRole:
    """Map authenticated user to executive dashboard persona."""
    if user.is_superuser:
        return "admin"

    role_names = {role.name for role in user.roles}
    if role_names & {"super_admin", "admin"}:
        return "admin"
    if "manager" in role_names:
        return "manager"
    if "employee" in role_names:
        return "employee"

    if user_has_permission(user, "audit.view") and user_has_permission(
        user, "transaction.view"
    ):
        return "admin"
    if user_has_permission(user, "workflow.approve"):
        return "manager"
    return "employee"


def _display_name(user: User) -> str:
    parts = [p for p in (user.first_name, user.last_name) if p]
    return " ".join(parts) if parts else user.email


class ExecutiveDashboardService:
    """Role-aware executive dashboard aggregation."""

    def __init__(
        self,
        audit_service: AuditService,
        notification_service: NotificationService,
        customer_repo: CustomerRepository,
        user_repo: UserRepository,
        workflow_repo: WorkflowRepository,
        report_repo: ReportRepository,
        app_metrics_repo: AppMetricsRepository,
        digital_inventory_repo: DigitalMetalInventoryRepository,
        metal_price_service: MetalPriceService,
        inventory_service: Optional[InventoryService] = None,
        transaction_service: Optional[TransactionService] = None,
    ):
        self.audit_service = audit_service
        self.notification_service = notification_service
        self.customer_repo = customer_repo
        self.user_repo = user_repo
        self.workflow_repo = workflow_repo
        self.report_repo = report_repo
        self.app_metrics_repo = app_metrics_repo
        self.digital_inventory_repo = digital_inventory_repo
        self.metal_price_service = metal_price_service
        self.inventory_service = inventory_service
        self.transaction_service = transaction_service

    async def get_dashboard(self, user: User) -> ExecutiveDashboardResponse:
        role = resolve_executive_role(user)
        cache_key = f"{user.id}:{role}"
        now_mono = time.monotonic()
        cached = _executive_cache.get(cache_key)
        if cached and (now_mono - cached[0]) < settings.DASHBOARD_CACHE_TTL_SECONDS:
            return cached[1]

        unread = await self.notification_service.get_unread_count(user.id)
        refreshed_at = datetime.now(timezone.utc)

        if role == "admin":
            payload = await self._build_admin_dashboard(user, unread, refreshed_at)
        elif role == "manager":
            payload = await self._build_manager_dashboard(user, unread, refreshed_at)
        else:
            payload = await self._build_employee_dashboard(user, unread, refreshed_at)

        _executive_cache[cache_key] = (now_mono, payload)
        return payload

    async def _build_admin_dashboard(
        self, user: User, unread: int, refreshed_at: datetime
    ) -> ExecutiveDashboardResponse:
        activity_trend: list = []
        if user_has_permission(user, "audit.view"):
            activity_trend = await self.audit_service.get_activity_trend(
                days=7, user_id=None
            )

        revenue_trend: list[RevenueTrendPoint] = []
        revenue_growth = None
        transaction_metrics = None
        customer_metrics = None
        inventory_metrics = None
        app_metrics = None

        can_view_wallet = user_has_permission(user, "wallet.view")
        can_view_transactions = user_has_permission(user, "transaction.view")

        if can_view_wallet or can_view_transactions:
            now = datetime.now(timezone.utc)
            day_start = now.replace(hour=0, minute=0, second=0, microsecond=0)
            day_end = day_start.replace(hour=23, minute=59, second=59, microsecond=999999)
            month_start = day_start.replace(day=1)

            (
                total_revenue,
                monthly_revenue,
                daily_revenue,
                total_transactions,
                monthly_transactions,
                member_count,
                members_new,
                trend_rows,
                growth,
            ) = await asyncio.gather(
                self.app_metrics_repo.paid_revenue_sum(),
                self.app_metrics_repo.paid_revenue_sum(start=month_start, end=day_end),
                self.app_metrics_repo.paid_revenue_sum(start=day_start, end=day_end),
                self.app_metrics_repo.count_wallet_transactions(),
                self.app_metrics_repo.count_wallet_transactions(
                    start=month_start, end=day_end
                ),
                self.app_metrics_repo.count_app_members(),
                self.app_metrics_repo.count_new_members_this_month(),
                self.app_metrics_repo.payment_revenue_trend(days=30),
                self.app_metrics_repo.payment_revenue_growth_percent(),
            )

            metal_inventory_value = Decimal("0")
            gold_available = Decimal("0")
            silver_available = Decimal("0")
            low_stock_metal_count = 0

            if user_has_permission(user, "inventory.view"):
                metals = await self.digital_inventory_repo.list_all()
                prices = await self.metal_price_service.get_prices()
                price_by_metal = {
                    "gold": prices.gold.retail_price,
                    "silver": prices.silver.retail_price,
                }
                for row in metals:
                    available = row.available_weight_grams
                    rate = price_by_metal.get(row.metal_type, Decimal("0"))
                    metal_inventory_value += available * rate
                    if row.metal_type == "gold":
                        gold_available = available
                    elif row.metal_type == "silver":
                        silver_available = available
                    status = compute_stock_status(
                        available, row.low_stock_threshold_grams
                    )
                    if status in {"low_stock", "out_of_stock"}:
                        low_stock_metal_count += 1

            app_metrics = AppDashboardMetrics(
                total_revenue=total_revenue,
                monthly_revenue=monthly_revenue,
                daily_revenue=daily_revenue,
                total_transactions=total_transactions,
                monthly_transactions=monthly_transactions,
                member_count=member_count,
                members_new_this_month=members_new,
                metal_inventory_value=metal_inventory_value,
                gold_available_grams=gold_available,
                silver_available_grams=silver_available,
                low_stock_metal_count=low_stock_metal_count,
            )
            revenue_trend = [
                RevenueTrendPoint(
                    label=row["label"],
                    revenue=row["revenue"],
                    transaction_count=row.get("transaction_count", 0),
                )
                for row in trend_rows
            ]
            revenue_growth = growth

        if user_has_permission(user, "customer.view"):
            customer_metrics = CustomerDashboardMetrics(
                **(await self.customer_repo.dashboard_metrics())
            )

        if self.inventory_service and user_has_permission(user, "inventory.view"):
            inv = await self.inventory_service.get_metrics(low_stock_limit=5)
            inventory_metrics = InventoryDashboardMetrics(
                total_stock=inv.total_stock,
                inventory_value=inv.inventory_value,
                low_stock_count=inv.low_stock_count,
                low_stock_items=inv.low_stock_items,
            )

        return ExecutiveDashboardResponse(
            role="admin",
            display_name=_display_name(user),
            unread_notifications=unread,
            refreshed_at=refreshed_at,
            revenue_trend=revenue_trend,
            revenue_growth_percent=revenue_growth,
            customer_metrics=customer_metrics,
            app_metrics=app_metrics,
            inventory_metrics=inventory_metrics,
            transaction_metrics=transaction_metrics,
            activity_trend=activity_trend,
        )

    async def _build_manager_dashboard(
        self, user: User, unread: int, refreshed_at: datetime
    ) -> ExecutiveDashboardResponse:
        today_start = datetime.now(timezone.utc).replace(
            hour=0, minute=0, second=0, microsecond=0
        )
        can_view_audit = user_has_permission(user, "audit.view")
        can_view_users = user_has_permission(user, "user.view")
        can_approve = user_has_permission(user, "workflow.approve")

        login_stats = {"today": 0, "week": 0, "month": 0}
        activity_total = 0
        active_users = 0
        pending_count = 0
        pending_items: list = []

        if can_view_audit:
            login_stats = await self.audit_service.get_login_statistics(
                user_id=user.id, system_wide=True
            )
            _, activity_total = await self.audit_service.list_audit_logs(
                skip=0,
                limit=1,
                user_id=None,
                start_date=today_start,
            )

        if can_view_users:
            active_users = await self.user_repo.count_active_users()

        if can_approve:
            pending_count = await self.workflow_repo.count_pending()
            pending_items = await self.workflow_repo.list_filtered(
                skip=0,
                limit=8,
                state="pending",
                sort_by="pending_since",
                sort_order="asc",
            )

        inventory_alerts: list[InventoryItemResponse] = []
        if self.inventory_service and user_has_permission(user, "inventory.view"):
            inv = await self.inventory_service.get_metrics(low_stock_limit=10)
            inventory_alerts = inv.low_stock_items

        pending_approvals = [
            WorkflowApprovalSummary(
                id=item.id,
                request_number=item.request_number,
                title=item.title,
                state=item.state,
                requester_name=(
                    f"{item.requester.first_name or ''} {item.requester.last_name or ''}".strip()
                    if item.requester
                    else None
                )
                or (item.requester.email if item.requester else None),
                assignee_name=(
                    f"{item.assignee.first_name or ''} {item.assignee.last_name or ''}".strip()
                    if item.assignee
                    else None
                )
                or (item.assignee.email if item.assignee else None),
                pending_since=item.pending_since,
                escalation_level=item.escalation_level,
            )
            for item in pending_items
        ]

        team_metrics = TeamDashboardMetrics(
            active_users=active_users,
            pending_approvals=pending_count,
            logins_today=login_stats["today"],
            team_activity_today=activity_total,
        )

        return ExecutiveDashboardResponse(
            role="manager",
            display_name=_display_name(user),
            unread_notifications=unread,
            refreshed_at=refreshed_at,
            team_metrics=team_metrics,
            pending_approvals=pending_approvals,
            inventory_alerts=inventory_alerts,
        )

    async def _build_employee_dashboard(
        self, user: User, unread: int, refreshed_at: datetime
    ) -> ExecutiveDashboardResponse:
        assigned_coro = self.workflow_repo.list_filtered(
            skip=0,
            limit=10,
            assignee_id=user.id,
            state="pending",
            sort_by="pending_since",
            sort_order="asc",
        )
        my_requests_coro = self.workflow_repo.list_filtered(
            skip=0,
            limit=5,
            requester_id=user.id,
            sort_by="created_at",
            sort_order="desc",
        )
        activity_coro = self.audit_service.list_audit_logs(
            skip=0, limit=12, user_id=user.id
        )
        trend_coro = self.audit_service.get_activity_trend(days=7, user_id=user.id)

        (
            assigned_items,
            my_requests,
            activity_result,
            activity_trend,
        ) = await asyncio.gather(
            assigned_coro,
            my_requests_coro,
            activity_coro,
            trend_coro,
        )
        activity_logs, _ = activity_result

        assigned_tasks = [
            AssignedTaskSummary(
                id=item.id,
                request_number=item.request_number,
                title=item.title,
                state=item.state,
                request_type=item.request_type,
                submitted_at=item.submitted_at,
            )
            for item in assigned_items
        ]

        for draft in my_requests:
            if draft.state == "draft" and len(assigned_tasks) < 10:
                assigned_tasks.append(
                    AssignedTaskSummary(
                        id=draft.id,
                        request_number=draft.request_number,
                        title=draft.title,
                        state=draft.state,
                        request_type=draft.request_type,
                        submitted_at=draft.submitted_at,
                    )
                )

        daily_activities = [
            DailyActivityItem(
                id=log.id,
                action=log.action,
                entity_type=log.entity_type,
                entity_id=log.entity_id,
                timestamp=log.timestamp,
                description=self._activity_description(log.action, log.entity_type),
            )
            for log in activity_logs
        ]

        return ExecutiveDashboardResponse(
            role="employee",
            display_name=_display_name(user),
            unread_notifications=unread,
            refreshed_at=refreshed_at,
            assigned_tasks=assigned_tasks,
            daily_activities=daily_activities,
            activity_trend=activity_trend,
        )

    @staticmethod
    def _activity_description(action: str, entity_type: Optional[str]) -> str:
        readable = action.replace("_", " ").title()
        if entity_type:
            return f"{readable} on {entity_type}"
        return readable

File: frontend/lib/features/dashboard/domain/executive_dashboard.dart
import 'package:ags_gold/features/dashboard/domain/dashboard_stats.dart';
import 'package:ags_gold/features/inventory/domain/inventory_item.dart';
import 'package:ags_gold/features/transactions/domain/transaction.dart';

class RevenueTrendPoint {
  final String label;
  final double revenue;
  final int transactionCount;

  const RevenueTrendPoint({
    required this.label,
    required this.revenue,
    this.transactionCount = 0,
  });

  factory RevenueTrendPoint.fromJson(Map<String, dynamic> json) {
    return RevenueTrendPoint(
      label: json['label'] as String? ?? '',
      revenue: _parseDecimal(json['revenue']),
      transactionCount: json['transaction_count'] as int? ?? 0,
    );
  }
}

class AppDashboardMetrics {
  final double totalRevenue;
  final double monthlyRevenue;
  final double dailyRevenue;
  final int totalTransactions;
  final int monthlyTransactions;
  final int memberCount;
  final int membersNewThisMonth;
  final double metalInventoryValue;
  final double goldAvailableGrams;
  final double silverAvailableGrams;
  final int lowStockMetalCount;
  final int pendingSellRequests;
  final int sellRequestsThisMonth;

  const AppDashboardMetrics({
    required this.totalRevenue,
    required this.monthlyRevenue,
    required this.dailyRevenue,
    required this.totalTransactions,
    required this.monthlyTransactions,
    required this.memberCount,
    required this.membersNewThisMonth,
    required this.metalInventoryValue,
    required this.goldAvailableGrams,
    required this.silverAvailableGrams,
    this.lowStockMetalCount = 0,
    this.pendingSellRequests = 0,
    this.sellRequestsThisMonth = 0,
  });

  factory AppDashboardMetrics.fromJson(Map<String, dynamic> json) {
    return AppDashboardMetrics(
      totalRevenue: _parseDecimal(json['total_revenue']),
      monthlyRevenue: _parseDecimal(json['monthly_revenue']),
      dailyRevenue: _parseDecimal(json['daily_revenue']),
      totalTransactions: json['total_transactions'] as int? ?? 0,
      monthlyTransactions: json['monthly_transactions'] as int? ?? 0,
      memberCount: json['member_count'] as int? ?? 0,
      membersNewThisMonth: json['members_new_this_month'] as int? ?? 0,
      metalInventoryValue: _parseDecimal(json['metal_inventory_value']),
      goldAvailableGrams: _parseDecimal(json['gold_available_grams']),
      silverAvailableGrams: _parseDecimal(json['silver_available_grams']),
      lowStockMetalCount: json['low_stock_metal_count'] as int? ?? 0,
      pendingSellRequests: json['pending_sell_requests'] as int? ?? 0,
      sellRequestsThisMonth: json['sell_requests_this_month'] as int? ?? 0,
    );
  }
}

class CustomerDashboardMetrics {
  final int totalCustomers;
  final int activeCustomers;
  final int newThisMonth;

  const CustomerDashboardMetrics({
    required this.totalCustomers,
    required this.activeCustomers,
    required this.newThisMonth,
  });

  factory CustomerDashboardMetrics.fromJson(Map<String, dynamic> json) {
    return CustomerDashboardMetrics(
      totalCustomers: json['total_customers'] as int? ?? 0,
      activeCustomers: json['active_customers'] as int? ?? 0,
      newThisMonth: json['new_this_month'] as int? ?? 0,
    );
  }
}

class TeamDashboardMetrics {
  final int activeUsers;
  final int pendingApprovals;
  final int loginsToday;
  final int teamActivityToday;

  const TeamDashboardMetrics({
    required this.activeUsers,
    required this.pendingApprovals,
    required this.loginsToday,
    required this.teamActivityToday,
  });

  factory TeamDashboardMetrics.fromJson(Map<String, dynamic> json) {
    return TeamDashboardMetrics(
      activeUsers: json['active_users'] as int? ?? 0,
      pendingApprovals: json['pending_approvals'] as int? ?? 0,
      loginsToday: json['logins_today'] as int? ?? 0,
      teamActivityToday: json['team_activity_today'] as int? ?? 0,
    );
  }
}

class WorkflowApprovalSummary {
  final String id;
  final String requestNumber;
  final String title;
  final String state;
  final String? requesterName;
  final String? assigneeName;
  final DateTime? pendingSince;
  final int escalationLevel;

  const WorkflowApprovalSummary({
    required this.id,
    required this.requestNumber,
    required this.title,
    required this.state,
    this.requesterName,
    this.assigneeName,
    this.pendingSince,
    this.escalationLevel = 0,
  });

  factory WorkflowApprovalSummary.fromJson(Map<String, dynamic> json) {
    return WorkflowApprovalSummary(
      id: json['id'] as String,
      requestNumber: json['request_number'] as String? ?? '',
      title: json['title'] as String? ?? '',
      state: json['state'] as String? ?? '',
      requesterName: json['requester_name'] as String?,
      assigneeName: json['assignee_name'] as String?,
      pendingSince: json['pending_since'] != null
          ? DateTime.parse(json['pending_since'] as String)
          : null,
      escalationLevel: json['escalation_level'] as int? ?? 0,
    );
  }
}

class AssignedTaskSummary {
  final String id;
  final String requestNumber;
  final String title;
  final String state;
  final String requestType;
  final DateTime? submittedAt;

  const AssignedTaskSummary({
    required this.id,
    required this.requestNumber,
    required this.title,
    required this.state,
    required this.requestType,
    this.submittedAt,
  });

  factory AssignedTaskSummary.fromJson(Map<String, dynamic> json) {
    return AssignedTaskSummary(
      id: json['id'] as String,
      requestNumber: json['request_number'] as String? ?? '',
      title: json['title'] as String? ?? '',
      state: json['state'] as String? ?? '',
      requestType: json['request_type'] as String? ?? 'general',
      submittedAt: json['submitted_at'] != null
          ? DateTime.parse(json['submitted_at'] as String)
          : null,
    );
  }
}

class DailyActivityItem {
  final String id;
  final String action;
  final String? entityType;
  final String? entityId;
  final DateTime timestamp;
  final String description;

  const DailyActivityItem({
    required this.id,
    required this.action,
    this.entityType,
    this.entityId,
    required this.timestamp,
    required this.description,
  });

  factory DailyActivityItem.fromJson(Map<String, dynamic> json) {
    return DailyActivityItem(
      id: json['id'] as String,
      action: json['action'] as String? ?? '',
      entityType: json['entity_type'] as String?,
      entityId: json['entity_id'] as String?,
      timestamp: DateTime.parse(json['timestamp'] as String),
      description: json['description'] as String? ?? '',
    );
  }
}

class AdminPaymentItem {
  final String id;
  final String razorpayOrderId;
  final String? razorpayPaymentId;
  final String? bankRrn;
  final String? paymentMethod;
  final String? customerName;
  final String? customerMobile;
  final String? customerEmail;
  final String metal;
  final double grams;
  final double amountInr;
  final String status;
  final String? failureReason;
  final DateTime createdAt;
  final DateTime? paidAt;
  final double? gstPercent;
  final double? metalValueInr;
  final double? gstAmountInr;
  final double? razorpayFeeInr;
  final double? merchantSettlementInr;

  const AdminPaymentItem({
    required this.id,
    required this.razorpayOrderId,
    this.razorpayPaymentId,
    this.bankRrn,
    this.paymentMethod,
    this.customerName,
    this.customerMobile,
    this.customerEmail,
    required this.metal,
    required this.grams,
    required this.amountInr,
    required this.status,
    this.failureReason,
    required this.createdAt,
    this.paidAt,
    this.gstPercent,
    this.metalValueInr,
    this.gstAmountInr,
    this.razorpayFeeInr,
    this.merchantSettlementInr,
  });

  factory AdminPaymentItem.fromJson(Map<String, dynamic> json) {
    return AdminPaymentItem(
      id: json['id'] as String? ?? '',
      razorpayOrderId: json['razorpay_order_id'] as String? ?? '',
      razorpayPaymentId: json['razorpay_payment_id'] as String?,
      bankRrn: json['bank_rrn'] as String?,
      paymentMethod: json['payment_method'] as String?,
      customerName: json['customer_name'] as String?,
      customerMobile: json['customer_mobile'] as String?,
      customerEmail: json['customer_email'] as String?,
      metal: json['metal'] as String? ?? 'gold',
      grams: _parseDecimal(json['grams']),
      amountInr: _parseDecimal(json['amount_inr']),
      status: json['status'] as String? ?? 'created',
      failureReason: json['failure_reason'] as String?,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : DateTime.now(),
      paidAt: json['paid_at'] != null
          ? DateTime.parse(json['paid_at'] as String)
          : null,
      gstPercent: json['gst_percent'] != null
          ? _parseDecimal(json['gst_percent'])
          : null,
      metalValueInr: json['metal_value_inr'] != null
          ? _parseDecimal(json['metal_value_inr'])
          : null,
      gstAmountInr: json['gst_amount_inr'] != null
          ? _parseDecimal(json['gst_amount_inr'])
          : null,
      razorpayFeeInr: json['razorpay_fee_inr'] != null
          ? _parseDecimal(json['razorpay_fee_inr'])
          : null,
      merchantSettlementInr: json['merchant_settlement_inr'] != null
          ? _parseDecimal(json['merchant_settlement_inr'])
          : null,
    );
  }
}

class AdminPaymentSummary {
  final double totalCapturedRevenue;
  final int totalCapturedCount;
  final int totalPendingCount;
  final int totalFailedCount;
  final double todayCapturedRevenue;
  final int todayCapturedCount;
  final DateTime? lastSyncedAt;

  const AdminPaymentSummary({
    required this.totalCapturedRevenue,
    required this.totalCapturedCount,
    required this.totalPendingCount,
    required this.totalFailedCount,
    required this.todayCapturedRevenue,
    required this.todayCapturedCount,
    this.lastSyncedAt,
  });

  factory AdminPaymentSummary.fromJson(Map<String, dynamic> json) {
    return AdminPaymentSummary(
      totalCapturedRevenue: _parseDecimal(json['total_captured_revenue']),
      totalCapturedCount: json['total_captured_count'] as int? ?? 0,
      totalPendingCount: json['total_pending_count'] as int? ?? 0,
      totalFailedCount: json['total_failed_count'] as int? ?? 0,
      todayCapturedRevenue: _parseDecimal(json['today_captured_revenue']),
      todayCapturedCount: json['today_captured_count'] as int? ?? 0,
      lastSyncedAt: json['last_synced_at'] != null
          ? DateTime.parse(json['last_synced_at'] as String)
          : null,
    );
  }
}

class CustomerPaymentSummary {
  final String mobile;
  final String? name;
  final String? email;
  final double totalPaidInr;
  final int successCount;
  final double totalGrams;
  final List<AdminPaymentItem> payments;

  const CustomerPaymentSummary({
    required this.mobile,
    this.name,
    this.email,
    required this.totalPaidInr,
    required this.successCount,
    this.totalGrams = 0.0,
    this.payments = const [],
  });

  factory CustomerPaymentSummary.fromJson(Map<String, dynamic> json) {
    return CustomerPaymentSummary(
      mobile: json['mobile'] as String? ?? 'Unknown',
      name: json['name'] as String?,
      email: json['email'] as String?,
      totalPaidInr: _parseDecimal(json['total_paid_inr']),
      successCount: json['success_count'] as int? ?? 0,
      totalGrams: _parseDecimal(json['total_grams']),
      payments: (json['payments'] as List<dynamic>?)
              ?.map((e) => AdminPaymentItem.fromJson(e as Map<String, dynamic>))
              .toList() ??
          const [],
    );
  }
}

class ExecutiveDashboard {
  final String role;
  final String displayName;
  final int unreadNotifications;
  final DateTime refreshedAt;
  final List<RevenueTrendPoint> revenueTrend;
  final double? revenueGrowthPercent;
  final CustomerDashboardMetrics? customerMetrics;
  final AppDashboardMetrics? appMetrics;
  final InventoryMetrics? inventoryMetrics;
  final TransactionMetrics? transactionMetrics;
  final TeamDashboardMetrics? teamMetrics;
  final List<WorkflowApprovalSummary> pendingApprovals;
  final List<InventoryItem> inventoryAlerts;
  final List<AssignedTaskSummary> assignedTasks;
  final List<DailyActivityItem> dailyActivities;
  final List<ActivityTrendPoint> activityTrend;
  final List<AdminPaymentItem> recentPayments;
  final AdminPaymentSummary? paymentSummary;
  final List<CustomerPaymentSummary> customerSummaries;

  const ExecutiveDashboard({
    required this.role,
    required this.displayName,
    required this.unreadNotifications,
    required this.refreshedAt,
    this.revenueTrend = const [],
    this.revenueGrowthPercent,
    this.customerMetrics,
    this.appMetrics,
    this.inventoryMetrics,
    this.transactionMetrics,
    this.teamMetrics,
    this.pendingApprovals = const [],
    this.inventoryAlerts = const [],
    this.assignedTasks = const [],
    this.dailyActivities = const [],
    this.activityTrend = const [],
    this.recentPayments = const [],
    this.paymentSummary,
    this.customerSummaries = const [],
  });

  ExecutiveDashboard copyWith({
    String? role,
    String? displayName,
    int? unreadNotifications,
    DateTime? refreshedAt,
    List<RevenueTrendPoint>? revenueTrend,
    double? revenueGrowthPercent,
    CustomerDashboardMetrics? customerMetrics,
    AppDashboardMetrics? appMetrics,
    InventoryMetrics? inventoryMetrics,
    TransactionMetrics? transactionMetrics,
    TeamDashboardMetrics? teamMetrics,
    List<WorkflowApprovalSummary>? pendingApprovals,
    List<InventoryItem>? inventoryAlerts,
    List<AssignedTaskSummary>? assignedTasks,
    List<DailyActivityItem>? dailyActivities,
    List<ActivityTrendPoint>? activityTrend,
    List<AdminPaymentItem>? recentPayments,
    AdminPaymentSummary? paymentSummary,
    List<CustomerPaymentSummary>? customerSummaries,
  }) {
    return ExecutiveDashboard(
      role: role ?? this.role,
      displayName: displayName ?? this.displayName,
      unreadNotifications: unreadNotifications ?? this.unreadNotifications,
      refreshedAt: refreshedAt ?? this.refreshedAt,
      revenueTrend: revenueTrend ?? this.revenueTrend,
      revenueGrowthPercent: revenueGrowthPercent ?? this.revenueGrowthPercent,
      customerMetrics: customerMetrics ?? this.customerMetrics,
      appMetrics: appMetrics ?? this.appMetrics,
      inventoryMetrics: inventoryMetrics ?? this.inventoryMetrics,
      transactionMetrics: transactionMetrics ?? this.transactionMetrics,
      teamMetrics: teamMetrics ?? this.teamMetrics,
      pendingApprovals: pendingApprovals ?? this.pendingApprovals,
      inventoryAlerts: inventoryAlerts ?? this.inventoryAlerts,
      assignedTasks: assignedTasks ?? this.assignedTasks,
      dailyActivities: dailyActivities ?? this.dailyActivities,
      activityTrend: activityTrend ?? this.activityTrend,
      recentPayments: recentPayments ?? this.recentPayments,
      paymentSummary: paymentSummary ?? this.paymentSummary,
      customerSummaries: customerSummaries ?? this.customerSummaries,
    );
  }

  factory ExecutiveDashboard.fromJson(Map<String, dynamic> json) {
    return ExecutiveDashboard(
      role: json['role'] as String? ?? 'employee',
      displayName: json['display_name'] as String? ?? '',
      unreadNotifications: json['unread_notifications'] as int? ?? 0,
      refreshedAt: DateTime.parse(json['refreshed_at'] as String),
      revenueTrend:
          (json['revenue_trend'] as List<dynamic>?)
              ?.map(
                (e) => RevenueTrendPoint.fromJson(e as Map<String, dynamic>),
              )
              .toList() ??
          const [],
      revenueGrowthPercent: json['revenue_growth_percent'] != null
          ? _parseDecimal(json['revenue_growth_percent'])
          : null,
      customerMetrics: json['customer_metrics'] != null
          ? CustomerDashboardMetrics.fromJson(
              json['customer_metrics'] as Map<String, dynamic>,
            )
          : null,
      appMetrics: json['app_metrics'] != null
          ? AppDashboardMetrics.fromJson(
              json['app_metrics'] as Map<String, dynamic>,
            )
          : null,
      inventoryMetrics: json['inventory_metrics'] != null
          ? InventoryMetrics.fromJson(
              json['inventory_metrics'] as Map<String, dynamic>,
            )
          : null,
      transactionMetrics: json['transaction_metrics'] != null
          ? TransactionMetrics.fromJson(
              json['transaction_metrics'] as Map<String, dynamic>,
            )
          : null,
      teamMetrics: json['team_metrics'] != null
          ? TeamDashboardMetrics.fromJson(
              json['team_metrics'] as Map<String, dynamic>,
            )
          : null,
      pendingApprovals:
          (json['pending_approvals'] as List<dynamic>?)
              ?.map(
                (e) =>
                    WorkflowApprovalSummary.fromJson(e as Map<String, dynamic>),
              )
              .toList() ??
          const [],
      inventoryAlerts:
          (json['inventory_alerts'] as List<dynamic>?)
              ?.map((e) => InventoryItem.fromJson(e as Map<String, dynamic>))
              .toList() ??
          const [],
      assignedTasks:
          (json['assigned_tasks'] as List<dynamic>?)
              ?.map(
                (e) => AssignedTaskSummary.fromJson(e as Map<String, dynamic>),
              )
              .toList() ??
          const [],
      dailyActivities:
          (json['daily_activities'] as List<dynamic>?)
              ?.map(
                (e) => DailyActivityItem.fromJson(e as Map<String, dynamic>),
              )
              .toList() ??
          const [],
      activityTrend:
          (json['activity_trend'] as List<dynamic>?)
              ?.map(
                (e) => ActivityTrendPoint.fromJson(e as Map<String, dynamic>),
              )
              .toList() ??
          const [],
      recentPayments:
          (json['recent_payments'] as List<dynamic>?)
              ?.map(
                (e) => AdminPaymentItem.fromJson(e as Map<String, dynamic>),
              )
              .toList() ??
          const [],
      paymentSummary: json['payment_summary'] != null
          ? AdminPaymentSummary.fromJson(
              json['payment_summary'] as Map<String, dynamic>,
            )
          : null,
      customerSummaries: (json['customer_summaries'] as List<dynamic>?)
              ?.map(
                (e) => CustomerPaymentSummary.fromJson(
                  e as Map<String, dynamic>,
                ),
              )
              .toList() ??
          const [],
    );
  }
}

double _parseDecimal(dynamic value) {
  if (value is num) return value.toDouble();
  if (value is String) return double.tryParse(value) ?? 0;
  return 0;
}
